kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -206,15 +206,17 @@ public <KOut, VOut> void scheduleForProcessing(
             // Once we've successfully placed a failed event in the finalizing queue
             // then there's nothing more to handle by the async thread pool since the
             // StreamThread will process the exception from here
-            return null;
           })
           .exceptionally(fatalException -> {
             // do this alone & in separate stage to ensure we always catch a fatal exception, even
             // if we somehow hit another exception while handling an exception in a previous stage
             fatalExceptions.computeIfAbsent(
                 inFlightKey,
                 k -> new FatalAsyncException("Uncaught exception while handling", fatalException));
-            return fatalException;
+            if (fatalException instanceof RuntimeException) {
+              throw (RuntimeException) fatalException;
+            }
+            throw new RuntimeException(fatalException);
```

Comment:
nit: extract the `new FatalAsyncException("Uncaught exception while handling", fatalException))` above and then just throw that here.

The main reason I added that was to help clean up all this RuntimeException casting and wrapping